docs/diplomat-and-macros.md

```diff
@@ -0,0 +1,272 @@
+
+```Rust
+pub struct MyFFIType {
+    pub a: i32,
+    pub b: bool,
+}
+
+#[diplomat::bridge]
+mod ffi {
+    impl MyFFIType {
+        pub fn create() -> MyFFIType { ... }
+        ...
+    }
+}
+```
+
+fails - diplomat can't handle this scenario - in the same way and for the same reasons that Ryan's
+[#416](https://github.com/mozilla/uniffi-rs/pull/416) can't - the contents of the struct aren't known.
+
+From the Rust side of the world, this is probably solvable by sprinkling more macros around - eg, something like:
+
+```Rust
+#[uniffi::magic]
+pub struct MyFFIType {
+    pub a: i32,
+    pub b: bool,
+}
+```
+
+Might be enough for the generation of the Rust scaffolding - in UniFFI's case, all we really need
+is an implementation of `uniffi::RustBufferViaFfi` which is easy to derive, and UniFFI can
+generate code which assumes that exists, much like it does now.
+However, the problems are in the foreign bindings, because, eg, those foreign bindings do not know
+the names and types of the struct elements without re-parsing every bit of Rust code with those
+annotations. As discussed below, re-parsing this code might be an option if we help Uniffi to
+find it, but asking UniFFI to parse this and all dependent crates to auto-discover them
+probably is not going to be viable.
+
+### Why is this considered a limition for UniFFI but not diplomat?
+
+As mentioned above, diplomat considers the limitation described above as an intentional design
+feature. By limiting where FFI types can be described, there's no risk of changes made "far away"
+from the FFI to change the FFI. This was born of experience in tools like `cbindgen`.
+
+For Uniffi, all use-cases needed by Mozilla don't share this design goal, primarily because the
+FFI is the primary consumer of the crate. The Rust API exists purely to service the FFI. It's not
+really possible to accidentally change the API, because every API change made will be in service
+of exposing that change over the FFI. The test suites written in the foreign languages are
+considered canonical.
+
+## How the type universe is constructed for the macro approach.
+
+In both diplomat and [#416](https://github.com/mozilla/uniffi-rs/pull/416), the approach taken
+is very similar - it takes a path to a the Rust source file/tree, and uses `syn` to locate the special modules (ie, ones annotated with `#[diplomat:bridge]` in the case of diplomat.)
+
+While some details differ, this is just a matter of implementation - #416 isn't quite as agressive
+about consuming the entire crate to find multiple FFI modules (and even then, diplomat doesn't
+actually *process* the entire crate, just modules tagged as a bridge), but could easily be
+extended to do so.
+
+But in both cases, for our problematic example above, this process never sees the layout of the
+`MyFFIType` struct, so that layout can't be communicated to the foreign bindings.
+As noted above, this is considered a feature for diplomat, but a limitation for UniFFI.
```

Comment:
How does this work in Diplomat, does the FFI allow you to pass it around as a pointer but not construct one for yourself on the foreign-language side?

Reply:
No - diplomat doesn't have that problem because the struct definition must appear inside the module, so the foreign bindings, which parse that module, do know the struct elements. This is the same as we discovered in #416 - that forcing all type definitions into the single ffi module would technically work, but the impact that would have on how our code is organized made it less appealing than the status quo.

I added a few words here to try and make that clearer.